specification/sections/01-flag-evaluation.md

```diff
@@ -31,6 +33,20 @@ OpenFeature.setProvider(new MyProvider());
 
 See [provider](./02-providers.md) for details.
 
+#### Requirement 1.1.2.2
+
+> The `provider mutator` function **MUST** invoke the `initialize` function on the newly registered provider before using it to resolve flag values.
```

Comment:
I tried to implement this in Java. It turns out that it's more complex than you'd expect b/c the Provider doesn't know if it's ready or not, per the spec. Only the global singleton knows.

This means the API will need to manage the lifecycle which is really awkward. I'd like it if we could add something to the spec which says "providers need to be able to tell us what it's current state is on demand via some accessor function".

Reply:
one option is to mandate a `provider state` which is exposed through the accessor. For example, `ready`, `pending` or `error` state, which can be used to react in evaluations or other usages 🤔 

And in case if the provider does not support init and shutdown, the default can be set to `ready`.

Reply:
> I tried to implement this in Java. It turns out that it's more complex than you'd expect b/c the Provider doesn't know if it's ready or not, per the spec. Only the global singleton knows.
> 
> This means the API will need to manage the lifecycle which is really awkward. I'd like it if we could add something to the spec which says "providers need to be able to tell us what it's current state is on demand via some accessor function".



> one option is to mandate a `provider state` which is exposed through the accessor. For example, `ready`, `pending` or `error` state, which can be used to react in evaluations or other usages thinking
> 
> And in case if the provider does not support init and shutdown, the default can be set to `ready`.

I think I agree. I actually implemented this initially, but then removed it for simplicity... but you're right, it means the API has to track it, which IS awkward. I think it makes sense to add this back to the provider.

Reply:
Boolean indicating whether it's ready seems sufficient to implement this.

Reply:
👍 for a boolean with a ready state

Reply:
OK I will add this.

Reply:
Added, see [this commit](https://github.com/open-feature/spec/pull/179/commits/164071b5a1ad19553c19156c8a0b68521a9324d9) for the delta. I hope it's clear enough.

@Kavindu-Dodan @justinabrahms

Reply:
Fine from my end :)

Reply:
@Kavindu-Dodan @justinabrahms @beeme1mr

I'm trying to decide exactly what makes more sense here... though it's simple, I'm somewhat unsure.

In the experimental web SDK, the SDK calls the `initialize()` function when the provider is set, and if that succeeds it updates it's own state tracking the provider's readiness (this is the awkwardness @justinabrahms referred to above). I see the point that maintaining this on the provider feels better, but what I'm not sure of is if the SDK should set that value or the provider should. Does it make more sense for the SDK to run the `initialize()` function and then set this field to ready? Or should we leave that field as read-only from outside and rely on the provider author to set it, and specify that the provider should set it during `initialize()`?

Should we specify something like `the intitialize() function will run if the "ready" state is set to false`? I just worry that this will be an easy thing for providers to screw up. I'm wondering if we should actually just maintain the state in the SDK itself after all.

Reply:
Generally it feels better, to have the state of the provider as a property of the provider.
I would totally agree that lowering the coherence of the provider by pulling provider state out of the provider is somewhat awkward.

>  I just worry that this will be an easy thing for providers to screw up. 

This could be mitigated by the SDK implementations. E.g. the default implementation of that method could just be setting the `ready state` to true, so providers would have to overwrite the method and call the super method.
Also It seems to me that specifying that the `ready flag` has to be set to true after running the initialize, should be enough to test providers for their spec compliance.

Reply:
What I was also wondering about is what happens if the client using that provider is doing feature flag evaluation before the `ready flag` is set to true.
I think there is nothing yet defined in the spec right? If thats true I think we should add that to the specification.
I think the spec should mention that in that case a PROVIDER_NOT_READY should be returned.

Reply:
> This could be mitigated by the SDK implementations. E.g. the default implementation of that method could just be setting the `ready flag` to true, so providers would have to overwrite the method and call the super method.

Not all languages have default implementations, but I think there's equivalents and I agree with your point in principle.

> Also It seems to me that specifying that the `ready flag` has to be set to true after running the initialize, should be enough to test providers for their spec compliance.

To be clear: do you feel the SDK should set this, or the provider should? Is this a requirement on the SDK or the provider author?

> I think the spec should mention that in that case a PROVIDER_NOT_READY should be returned.

I think this might be a good idea. Yes. I'll think about that a bit more.

Reply:
> Not all languages have default implementations, but I think there's equivalents and I agree with your point in principle.

Yeah thats true, I just had in mind that in many cases this at least weakens the problem.

> To be clear: do you feel the SDK should set this, or the provider should? Is this a requirement on the SDK or the provider author?

I would say it is a requirement on the provider author, or maybe better the provider implementation.
This also means that if the provider defines that flag, it should always be initially false, which is different from not defining the flag as the spec says that in this case the provider should be considered ready.

> I think this might be a good idea. Yes. I'll think about that a bit more.

Sounds good! 👍

Reply:
> This also means that if the provider defines that flag, it should always be initially false, which is different from not defining the flag as the spec says that in this case the provider should be considered ready.

I think there are providers that are "immediately ready" such as those that pull from environment variables or otherwise get their configuration synchronously from their environment.

Reply:
I think it is reasonable to fallback to the default value of the flag when the provider is not ready yet. And then pass the reason as suggested in the evaluation result.  Then the implementor/consumer of OpenFeature can decide whether the want to postpone execution until the provider is ready or fallback to the default value of the flag?

Reply:
@lukas-reining @weyert thanks!

I've made the changes described above. I've made the language clearer about the fact it's the provider's responsibility to set the ready member, and also added a point about resolution with providers that are not ready. See [delta](https://github.com/open-feature/spec/pull/179/commits/e5ba035ce2f256e8259dd9d9a009123c4d1733c6).

Reply:
Looks good, this makes it clearer🙂